### PR TITLE
Attach looks at future observatory URIs

### PR DESCRIPTION
**Pull request by blasten**
## Description

I refactored the the protocol discovery mechanism, so it looks at future observatory URIs instead of old ones.  I'm sending this PR to get early feedback. I will then add unit tests.

The main change is that `observatoryUri` becomes a `Stream` of `Uri`.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
Attach changes generally look good I think - though I would do more verification to ensure that we're cleaning exiting the loop on errors and cleaning up resources in between.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -124,7 +124,7 @@ class FlutterDevice {
 
   final Device device;
   final ResidentCompiler generator;
-  List<Uri> observatoryUris;
+  Stream<Uri> observatoryUris;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/resident_runner.dart]: What does changing this to a Stream do?  I think It takes a list so that for `-d all` the runner can handle multiple vmservice addresses

Comment (zanderso) [reply packages/flutter_tools/lib/src/resident_runner.dart]: If I'm reading this correctly, it looks like as a list this has either 0 or 1 elements, so I don't think it's related to -d all. What I'm not sure about is why it had to be an empty list rather than null when there were no observatory uris.

**Review by zanderso — COMMENTED** [review]
Instead of making `ProtocolDiscovery` give a stream of Uris, did you consider having the `attach` command instead do a loop around `_attachToDevice`?

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/resident_runner.dart]: It appears the intention was to support multiple apps running an observatory on a single device. Was this done for TQ?

**Comment by blasten** [comment]
Capturing the offline conversation:

The goal here is to discover the observatory URI of an app:
1.  Already running on the device.
2. That is restarted. (e.g. killed and then started again) 
3. That starts after running `flutter attach`. (Currently supported)

cc @zanderso

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 2]
```diff
@@ -321,16 +326,25 @@ class AttachCommand extends FlutterCommand {
...
-        unawaited(onAppStart.future.whenComplete(() {
-          TerminalHandler(runner)
-            ..setupTerminal()
-            ..registerSignalHandlers();
-        }));
-        result = await runner.attach(
-          appStartedCompleter: onAppStart,
-        );
-        assert(result != null);
+        while (true) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/attach.dart]: What happens on the second iteration of this loop when `ProtocolDiscovery` is not used above?

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Yeah, it looks like this was for the now-deleted fuchsia_reload command. Fuchsia is now handled differently, so I think this can be safely modified.

**Comment by zanderso** [comment]
Overall approach looks good. Need tests of reconnect, and follow-on PRs (or same PR, up to you) to support re-attaching with mdns (and other) discovery.

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 3]
```diff
@@ -247,7 +247,7 @@ class AttachCommand extends FlutterCommand {
             portForwarder: device.portForwarder,
           );
           printStatus('Waiting for a connection from Flutter on ${device.name}...');
-          observatoryUri = observatoryDiscovery.uris;
+          observatoryUri = Stream<Uri>.fromFuture(observatoryDiscovery.uris.first).asBroadcastStream();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/attach.dart]: Won't a broadcast stream have timing issues? Events aren't buffered, if the subscriber is added after discovery it would be missed

File: packages/flutter_tools/lib/src/commands/attach.dart (hunk 2 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/commands/attach.dart]: it now breaks if the stream is closed.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 1 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/resident_runner.dart]: As per offline discussion, I changed this such that it doesn't reuse the resident runner, which aligns with the expected lifecycle of the resident runner.

**Comment by blasten** [comment]
This PR is ready for a final review. PTAL

**Comment by xster** [comment]
cc @devoncarew, @DanTup this should make the IDE specific work-around for Android flutter attach redundant.

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 4]
```diff
@@ -49,50 +57,78 @@ class ProtocolDiscovery {
...
 
+  StreamController<Uri> _uriStreamController;
+
   /// The discovered service URI.
+  /// Use [uris] instead.
+  // TODO(egarciad): replace `uri` for `uris`.
+  Future<Uri> get uri async {
+    try {
+      return await uris.first;
+    } on StateError {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: This seems like a bug. We generally shouldn't be catching Error subclasses unless it is a temporary work around. If you are going to replace uri with uris I would do that in this PR

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 5]
```diff
@@ -49,50 +57,78 @@ class ProtocolDiscovery {
...
-    return await _forwardPort(rawUri);
+  ///
+  /// Dependending on the lifespan of the app running the observatory,
+  /// a new observatory URI may be assigned to the app.
+  Stream<Uri> get uris {
+    return _uriStreamController.stream
+      .transform(_throttle<Uri>(
+        timeInMilliseconds: throttleTimeInMilliseconds,
+      ))
+      .asyncMap<Uri>((Uri observatoryUri) async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: async/await is redundant:

```suggestion
      .asyncMap<Uri>((Uri observatoryUri) =>  _forwardPort(observatoryUri));
```

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 6]
```diff
@@ -218,8 +218,7 @@ class AttachCommand extends FlutterCommand {
         FuchsiaIsolateDiscoveryProtocol isolateDiscoveryProtocol;
         try {
           isolateDiscoveryProtocol = device.getIsolateDiscoveryProtocol(module);
-          observatoryUri = await isolateDiscoveryProtocol.uri;
-          printStatus('Done.'); // FYI, this message is used as a sentinel in tests.
+          observatoryUri = Stream<Uri>.fromFuture(isolateDiscoveryProtocol.uri).asBroadcastStream();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/attach.dart]: Here and elsewhere, don't use broadcast streams. These do not buffer events, so if the future resolves before the first subscriber is added the event will be lost

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 7]
```diff
@@ -154,23 +158,49 @@ class FlutterDevice {
...
-        observatoryUris[i],
-        reloadSources: reloadSources,
-        restart: restart,
-        compileExpression: compileExpression,
-      );
-      printTrace('Successfully connected to service protocol: ${observatoryUris[i]}');
-    }
-    vmServices = localVmServices;
-    device.getLogReader(app: package).connectedVMServices = vmServices;
+  }) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/resident_runner.dart]: It looks like you re-added the updates to resident runner. Were these changes necessary for some reason?

**Review by jonahwilliams — DISMISSED** [review]

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 4 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: I'm not sure I follow. If something called cancel too early, then this is the equivalent behavior in the current code.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 7 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/resident_runner.dart]: yes -- `observatoryUris` is a stream instead of a list.

File: packages/flutter_tools/lib/src/commands/attach.dart (hunk 6 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/commands/attach.dart]: This seems to work fine:

```dart
import 'dart:async';

void main() async {
  Stream<String> stream = Stream<String>.fromFuture(Future<String>.value('foo'))
      .asBroadcastStream();

  stream.listen((String value) {
    print(value);
  });
}

```
prints `foo`.

Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/attach.dart]: The docs for `asBroadcastStream()` [here](https://api.dartlang.org/stable/2.6.1/dart-async/Stream/asBroadcastStream.html):

> The returned stream will subscribe to this stream when its first subscriber is added, and will stay subscribed until this stream ends, or a callback cancels the subscription.

In general though, I don't know if it would be needed to pass `onListen` and `onCancel` callbacks to `asBroadcastStream()` to pause and resume the underlying stream when the number of subscribers hits zero.

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 8]
```diff
@@ -325,19 +294,33 @@ class AttachCommand extends FlutterCommand {
         result = await app.runner.waitForAppToFinish();
         assert(result != null);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/attach.dart]: Can you return here and remove the 'else' ?

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 9]
```diff
@@ -49,50 +57,76 @@ class ProtocolDiscovery {
...
+    }
+    return null;
+  }
+
+  void _handleLine(String line) {
+    Uri uri;
+    try {
+      uri = _getObservatoryUri(line);
+    } on FormatException catch(error, stackTrace) {
+      _uriStreamController.addError(error, stackTrace);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Note that if the listener on `_uriStreamController.stream` doesn't have an `onError` callback, then this error will be thrown into the containing zone.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 10]
```diff
@@ -154,23 +158,49 @@ class FlutterDevice {
...
+    _isStreamingNewObservatoryUri = true;
+
+    final Completer<void> completer = Completer<void>();
+    StreamSubscription<void> subscription;
+    bool isWaitingForVm = false;
+
+    subscription = observatoryUris.listen((Uri observatoryUri) async {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: I suspect this listen() call should have an onError callback

File: packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart [hunk 11]
```diff
@@ -77,17 +77,15 @@ void main() {
...
+          mockLogReader.addLine('Foo');
+          mockLogReader.addLine('Observatory listening on http://127.0.0.1:$devicePort');
           return mockLogReader;
         });
         testDeviceManager.addDevice(device);
         final Completer<void> completer = Completer<void>();
         final StreamSubscription<String> loggerSubscription = logger.stream.listen((String message) {
-          if (message == '[stdout] Done.') {
-            // The "Done." message is output by the AttachCommand when it's done.
+          if (message == '[verbose] Observatory URL on device: http://127.0.0.1:499') {
```
Comment (zanderso) [inline packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart]: 499 -> $devicePort

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 4 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: In what situation does `_uriStreamController?.close()` triggers the `StateError`? When you say `uris.first` after the `_uriStreamController?.close()`? Can you check `_uriStreamController.isClosed` or add an `onCancel` callback to `_uriStreamController` that records a bit?

**Comment by devoncarew** [comment]
> cc @devoncarew, @DanTup this should make the IDE specific work-around for Android flutter attach redundant.

cc @stevemessick 

From my understanding, the user will now be able to run flutter attach before or after the native app launch, but will want to do it before the Flutter view in the app starts up? Sounds like an overall better UX. I don't think IDEs will need to change (just documentation)?

File: packages/flutter_tools/test/general.shard/protocol_discovery_test.dart [hunk 12]
```diff
@@ -134,15 +140,91 @@ void main() {
...
+
+        final List<Uri> discoveredUris = <Uri>[];
+        discoverer.uris.listen((Uri uri) {
+          discoveredUris.add(uri);
+        });
+
+        logReader.addLine('I/flutter : Observatory listening on http://127.0.0.1:12346/PTwjm8Ii8qg=/');
+        logReader.addLine('I/flutter : Observatory listening on http://127.0.0.1:12345/PTwjm8Ii8qg=/');
+
+        await Future<void>.delayed(const Duration(milliseconds: kThrottleTimeInMilliseconds * 2));
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/protocol_discovery_test.dart]: Use `FakeAsync`

File: packages/flutter_tools/lib/src/commands/attach.dart (hunk 8 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/commands/attach.dart]: +1

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 9 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: I added the `onError` callback. Thanks.

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 4 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: The situation is that there's a test **returns non-null uri future** under `protocol_discovery_test.dart` that expects `uri` to return a non-null value. For some reason, this works currently even though the `uri` future never completes. I checked the documentation for `expect` and it says:
" If  [actual] is a [Future], the test won't complete until the future emits a value."  However, it doesn't seem true in this case:
```.dart
final Completer<void> _completer = Completer<Uri>();
expect(_completer.future, isNotNull);
```


For this reason, the test is failing when migrated to streams.  The stream is closed in the `tearDown` callback before a value is emitted.

**Comment by DanTup** [comment]
> From my understanding, the user will now be able to run flutter attach before or after the native app launch, but will want to do it before the Flutter view in the app starts up?

Isn't this how things already are today?

**Comment by blasten** [comment]
This change will decouple the app lifecycle from the tool. Users can run flutter attach at any time and since it will look for new observatory connections, users don’t need to rerun the command.

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 13]
```diff
@@ -110,7 +138,38 @@ class ProtocolDiscovery {
...
+      handleData: (S value, EventSink<S> sink) {
+        latestLine = value;
+
+        final int currentTime = DateTime.now().millisecondsSinceEpoch;
+        lastExecution ??= currentTime;
+        final int remainingTime = currentTime - lastExecution;
+        final int nextExecutionTime = remainingTime > timeInMilliseconds
+          ? 0
+          : timeInMilliseconds - remainingTime;
+        throttleFuture ??= Future<void>
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: This will create a separate Future that completes at the desired time for every log line. Futures that complete 'at the same time' aren't guaranteed to complete in any particular order, so the log messages might get re-arranged. That might not be a problem for your current use-case, but it might confuse someone who wants to try to use this for something else.

I'd suggest trying something simpler. Maybe:
```dart
List<S> latestEvents = <S>[];
...
handleData: (S value, EventSink<S> sink) {
  latestEvents.add(value);
  if (throttleFuture != null) {
    return;
  }
  throttleFuture = Future<void>
    .delayed(Duration(milliseconds: timeInMilliseconds))
    .whenComplete(() {
      sink.addAll(latestEvents);
      latestEvents.clear();
    });
}
```

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 14]
```diff
@@ -110,7 +138,38 @@ class ProtocolDiscovery {
     if (ipv6) {
       hostUri = hostUri.replace(host: InternetAddress.loopbackIPv6.host);
     }
-
     return hostUri;
   }
 }
+
+/// Throttles a stream by [timeInMilliseconds].
+StreamTransformer<S, S> _throttle<S>({
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Maybe `_batch` instead of `_throttle` ?

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 15]
```diff
@@ -110,7 +138,38 @@ class ProtocolDiscovery {
     if (ipv6) {
       hostUri = hostUri.replace(host: InternetAddress.loopbackIPv6.host);
     }
-
     return hostUri;
   }
 }
+
+/// Throttles a stream by [timeInMilliseconds].
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Please expand this comment a bit.

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 16]
```diff
@@ -110,7 +138,38 @@ class ProtocolDiscovery {
...
+StreamTransformer<S, S> _throttle<S>({
+  @required int timeInMilliseconds,
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Maybe make this a `Duration` and rename to `batchDuration`.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 17]
```diff
@@ -154,23 +158,51 @@ class FlutterDevice {
...
+        return;
+      }
+      printTrace('Successfully connected to service protocol: $observatoryUri');
+
+      vmServices = <VMService>[service];
+      device.getLogReader(app: package).connectedVMServices = vmServices;
+      completer.complete();
+      await subscription.cancel();
+    }, onError: (dynamic error) {
+      printTrace('Fail to handle observatory URI: $error');
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Should this do a `completer.completeError()` in some situations?

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 18]
```diff
@@ -131,16 +131,20 @@ class FlutterDevice {
...
   ApplicationPackage package;
   List<String> fileSystemRoots;
   String fileSystemScheme;
   StreamSubscription<String> _loggingSubscription;
+  bool _isStreamingNewObservatoryUri;
   final String viewFilter;
   final bool trackWidgetCreation;
 
+  /// Whether the stream [observatoryUris] is still open.
+  bool get isStreamingNewObservatoryUri => _isStreamingNewObservatoryUri ?? false;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Can this be private or `@visibleForTesting`?

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 19]
```diff
@@ -131,16 +131,20 @@ class FlutterDevice {
...
   final ResidentCompiler generator;
-  List<Uri> observatoryUris;
+  Stream<Uri> observatoryUris;
   List<VMService> vmServices;
   DevFS devFS;
   ApplicationPackage package;
   List<String> fileSystemRoots;
   String fileSystemScheme;
   StreamSubscription<String> _loggingSubscription;
+  bool _isStreamingNewObservatoryUri;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Maybe `_isListeningForObservatoryUri` ?

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 20]
```diff
@@ -221,6 +253,7 @@ class FlutterDevice {
     if (flutterViews.any((FlutterView view) {
       return view != null &&
              view.uiIsolate != null &&
+             view.uiIsolate.pauseEvent != null &&
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Is this a separate bug fix?

Comment (blasten) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Yep. I hit this issue while debugging.

Comment (blasten) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Yep. I saw this issue while debugging.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 18 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/resident_runner.dart]: This is used from the attach.dart to exit the loop. I have renamed it to `isWaitingForObservatory` though.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 17 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/resident_runner.dart]: fwiw, the only case where this could happen is if `FormatException` is thrown while parsing the observatory URI.  I'm not sure how that would be possible and if it will happen consistently.

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 14 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: discussed offline

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 21]
```diff
@@ -110,7 +138,38 @@ class ProtocolDiscovery {
...
+}) {
+  assert(timeInMilliseconds != null);
+
+  S latestLine;
+  int lastExecution;
+  Future<void> throttleFuture;
```
[the remark this answers is not in the document]
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: discussed offline

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 22]
```diff
@@ -110,7 +138,43 @@ class ProtocolDiscovery {
...
+
+  return StreamTransformer<S, S>
+    .fromHandlers(
+      handleData: (S value, EventSink<S> sink) {
+        latestLine = value;
+
+        final int currentTime = DateTime.now().millisecondsSinceEpoch;
+        lastExecution ??= currentTime;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Should `lastExecution` be initialized earlier? The way this is written it looks like the first event will always be emitted no earlier than `throttleDuration` ms after the first event comes in. Is that what you had in mind?

File: packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart [hunk 23]
```diff
@@ -104,13 +123,116 @@ void main() {
...
+      });
+
       testUsingContext('Fails with tool exit on bad Observatory uri', () async {
         when(device.getLogReader()).thenAnswer((_) {
           // Now that the reader is used, start writing messages to it.
-          Timer.run(() {
-            mockLogReader.addLine('Foo');
-            mockLogReader.addLine('Observatory listening on http:/:/127.0.0.1:$devicePort');
-          });
+          mockLogReader.addLine('Foo');
```
Comment (zanderso) [inline packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart]: Why do some tests put this in `Timer.run()` and some don't? Also why is the `mockLogReader` disposed here, but not in other places?

Comment (blasten) [reply packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart]: Removed the last one in attach_test.dart and protocol_discovery_test.dart.  `Timer.run` isn't needed because the throttle transformer already delays the completion of the `Future` more than one microtask.

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 22 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: I tried it, but I got very different results since the initial value is time sensitive. The time spent between the initialization and the first event was in some cases greater than the actual wait time.

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 24]
```diff
@@ -49,50 +57,70 @@ class ProtocolDiscovery {
...
-    return await _forwardPort(rawUri);
+  ///
+  /// Dependending on the lifespan of the app running the observatory,
+  /// a new observatory URI may be assigned to the app.
+  Stream<Uri> get uris {
+    return _uriStreamController.stream
+      .transform(_throttle<Uri>(
+        waitDuration: throttleDuration,
+      ))
+      .asyncMap<Uri>((Uri observatoryUri) => _forwardPort(observatoryUri));
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Further golfing using a tear-off:

```suggestion
      .asyncMap<Uri>(_forwardPort);
```

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 25]
```diff
@@ -49,50 +57,70 @@ class ProtocolDiscovery {
...
+  Uri _getObservatoryUri(String line) {
+    final Match match = _getPatternMatch(line);
     if (match != null) {
-      try {
-        uri = Uri.parse(match[1]);
-      } on FormatException catch (error, stackTrace) {
-        _stopScrapingLogs();
-        _completer.completeError(error, stackTrace);
-      }
+      return Uri.parse(match[1]);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: Instead of catching the FormatError below, you can get the same behavior as the non-matching case by using `Uri.tryParse` which returns `null` on failures

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 26]
```diff
@@ -154,23 +158,50 @@ class FlutterDevice {
...
+      VMService service;
+
+      try {
+        service = await VMService.connect(
+          observatoryUri,
+          reloadSources: reloadSources,
+          restart: restart,
+          compileExpression: compileExpression,
+        );
+      } catch (exception) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/resident_runner.dart]: use on `on  Exception catch (exception)` to avoid catching TypeError or NoSuchMethodError, et cetera

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 25 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: Currently, the `Future` completes with the associated error object and stack trace. It might be the case that this isn't needed anymore, but I'd prefer to make behavioral changes like this in a separate PR.

File: packages/flutter_tools/lib/src/devfs.dart [hunk 27]
```diff
@@ -263,6 +264,13 @@ class DevFSException implements Exception {
   final StackTrace stackTrace;
 }
 
+HttpClient get httpClient {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/devfs.dart]: Can this be scoped down from a top-level public? If this is only used to initialize `_DevFSHttpWriter._client`, maybe it can be scoped down to the initializer for that field or to the constructor of `_DevFSHttpWriter`.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 17 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Yeah, that showed up in crash logging at some point, which is why the existing code tries to defend against it.

File: packages/flutter_tools/lib/src/devfs.dart (hunk 27 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/devfs.dart]: Done

**Review by zanderso — APPROVED** [review]

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
(If LG to @zanderso )

File: packages/flutter_tools/lib/src/protocol_discovery.dart [hunk 28]
```diff
@@ -49,50 +57,70 @@ class ProtocolDiscovery {
...
-  /// need to port forward.
-  Future<Uri> get uri async {
-    final Uri rawUri = await _completer.future;
-    return await _forwardPort(rawUri);
+  /// When a new observatory URI is available in [logReader],
+  /// the URIs are forwarded at most once every [throttleDuration].
+  ///
+  /// Port forwarding is only attempted when this is invoked,
+  /// for each observatory URI in the stream.
+  Stream<Uri> get uris {
```
Comment (gspencergoog) [inline packages/flutter_tools/lib/src/protocol_discovery.dart]: If you're changing this API's name anyhow, @Hixie  [has pointed out](https://github.com/flutter/flutter/pull/45200#discussion_r348794795) that there is no such thing as a URI, they're all URLs, so maybe this should be called `urls` instead?

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 29]
```diff
@@ -229,85 +228,55 @@ class AttachCommand extends FlutterCommand {
...
+          .fromFuture(
+            MDnsObservatoryDiscovery.instance.getObservatoryUri(
+              appId,
+              device,
+              usesIpv6: usesIpv6,
+              deviceVmservicePort: deviceVmservicePort,
+            )
+          ).asBroadcastStream();
       }
       // If MDNS discovery fails or we're not on iOS, fallback to ProtocolDiscovery.
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/attach.dart]: Sorry for not catching this earlier, but this fallback will no longer work since the Stream.fromFuture isn't going to return null. That is the null check below is no longer testing the success of the mdns lookup. Could you please do a small follow-up change that checks the return of `MDnsObservatoryDiscovery.instance.getObservatoryUri()` for null, and if it is null then uses `ProtocolDiscovery.observatory()` instead? Thanks!

Comment (blasten) [reply packages/flutter_tools/lib/src/commands/attach.dart]: Sounds good. I will follow up on that!

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 28 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: That makes sense. Thanks!

File: packages/flutter_tools/lib/src/protocol_discovery.dart (hunk 4 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/protocol_discovery.dart]: catching a StateError is definitely a bug. It implies there's some logic error in our code.
